Animation editor THO-146

DHazardFenrir opened the pull request:
Animation Editor with Skinning.
How to Test:
Create New Scene
Import Player o Zombunny to scene and load the respective model. 
Select the rootGameobject y select the animation Component. There are the controllers for play, pause, etc. If you click play the animation should show, stop should stop it. 

I erased the animation editor since it made the flow of playing, selecting gameobjects kind of cumbersome, so with everything in the component is has more easy access.

Note: There is some errour in the animation resource or the import, i dont know what is the error exactly, but some channels are not importing, causing that there aren't those channels for the bones, so the animations are not quite complete in a way. 

Functions to add since this feature is urgent and I cannot spend more time on it: If you delete the animation component and then you add it again manually, the animation component doesn't automatically search for the animation, so it stays empty.  It has possible performance problem in the getTransform method since it has two while loops (one for rotation and position) when looking through indexes.

DHazardFenrir commented:
> Shouldn't be merged into VS2 instead of main?

Its already merge with the stuff in VS2 not main

SirKaza commented on SobrassadaEngine/FileSystem/Importers/ModelImporter.cpp:
```diff
@@ -7,7 +7,9 @@
 #include "Model.h"
 #include "ProjectModule.h"
-#include "ResourceModel.h"
+#include "ResourceManagement/Resources/ResourceModel.h"
```
ResourceManagement/Resources/ not needed, take a look at includes directory of project

SirKaza commented on SobrassadaEngine/FileSystem/Importers/ModelImporter.cpp:
```diff
@@ -101,7 +103,52 @@ namespace ModelImporter
 
         newModel.SetUID(finalModelUID);
+        const std::string animName = FileSystem::GetFileNameWithoutExtension(filePath);
+        std::string animAssetPath       = ANIMATIONS_ASSETS_PATH + animName + ANIMATION_EXTENSION;
```
Not being used

SirKaza commented on SobrassadaEngine/FileSystem/Importers/ModelImporter.cpp:
```diff
@@ -101,7 +103,52 @@ namespace ModelImporter
         assetPath = targetFilePath + assetPath;
 
         newModel.SetUID(finalModelUID);
+        const std::string animName = FileSystem::GetFileNameWithoutExtension(filePath);
```
Not being used

SirKaza commented on SobrassadaEngine/FileSystem/Model.h:
```diff
@@ -31,13 +31,19 @@ class Model
 
     const std::vector<std::vector<NodeData>>& GetNodes() const { return nodes; }
     const Skin& GetSkin(int skinIndex) const { return skins[skinIndex]; }
+    UID GetAnimationUID() const { return animUID; }
```
const?

SirKaza commented on SobrassadaEngine/ResourceManagement/Resources/ResourceAnimation.h:
```diff
@@ -0,0 +1,33 @@
+#include "Resource.h"
+#include "Math/float3.h"
+#include "Math/Quat.h"
+#include <memory>
```
memory not needed at least here

SirKaza commented on SobrassadaEngine/ResourceManagement/Resources/ResourceAnimation.h:
```diff
@@ -0,0 +1,33 @@
+class ResourceAnimation : public Resource
+{
+  public:
+    std::unordered_map<std::string, Channel> channels;
```
Public variables after and before private variables (don't merge methods with variables)

SirKaza commented on SobrassadaEngine/ResourceManagement/Resources/ResourceAnimation.cpp:
```diff
@@ -0,0 +1,57 @@
+#include "ResourceAnimation.h"
+
+#include "Application.h"
+#include "DirectXTex/DirectXTex.h"
```
directxtex and imgui in the bottom group

SirKaza commented on SobrassadaEngine/Scene/Components/ComponentUtils.h:
```diff
@@ -23,6 +23,7 @@ enum ComponentType
     COMPONENT_CANVAS,
     COMPONENT_LABEL,
     COMPONENT_CAMERA,
+    COMPONENT_ANIMATION,
```
When adding new things, do it at the end, so current scenes are not broken

SirKaza commented on SobrassadaEngine/Scene/Components/Standalone/AnimController.h:
```diff
@@ -0,0 +1,41 @@
+    void Pause();
+    void Resume();
+    
+   ResourceAnimation* GetCurrentAnimation() { return animation; }
```
getters first then setters

SirKaza commented on SobrassadaEngine/Scene/Components/Standalone/AnimController.h:
```diff
@@ -0,0 +1,41 @@
+  
+   
+  private:
+    Quat Interpolate(Quat& first, Quat& second, float lambda);
```
This private method separate it from variables

SirKaza commented on SobrassadaEngine/Scene/Components/Standalone/AnimController.cpp:
```diff
@@ -0,0 +1,218 @@
+#include "AnimController.h"
```
After the header of this cpp, separation line to the rest

SirKaza commented on SobrassadaEngine/Scene/Components/Standalone/AnimController.cpp:
```diff
@@ -0,0 +1,218 @@
+AnimController::~AnimController()
+{
+   
+    if (animation != nullptr)
```
This if not needed, same with animation = nullptr right below

SirKaza commented on SobrassadaEngine/Scene/Components/Standalone/AnimController.cpp:
```diff
@@ -0,0 +1,218 @@
+                animation = nullptr;
+            }
+
+            return UPDATE_CONTINUE;
```
This return is needed?

SirKaza commented on SobrassadaEngine/Scene/Components/Standalone/AnimController.cpp:
```diff
@@ -0,0 +1,218 @@
+            }
+            else
+            {
+                float startTime = animChannel->posTimeStamps[prevIndex];
```
this 3 variables can be const

SirKaza commented on SobrassadaEngine/Scene/Components/Standalone/AnimController.cpp:
```diff
@@ -0,0 +1,218 @@
+            }
+            else
+            {
+                float startTime = animChannel->rotTimeStamps[prevIndex];
```
same here

SirKaza commented on SobrassadaEngine/Scene/Components/Standalone/AnimationComponent.h:
```diff
@@ -0,0 +1,52 @@
+    AnimationComponent(UID uid, GameObject* parent);
+    AnimationComponent(const rapidjson::Value& initialState, GameObject* parent);
+    
+    ~AnimationComponent();
```
override

SirKaza commented on SobrassadaEngine/Scene/Components/Standalone/AnimationComponent.h:
```diff
@@ -0,0 +1,52 @@
+    void OnResume();
+    void OnInspector();
+
+    void Clone(const Component* other) override;
```
constructors -> override -> normal methogs -> getters and setters

SirKaza commented on SobrassadaEngine/Scene/Components/Standalone/AnimationComponent.h:
```diff
@@ -0,0 +1,52 @@
+    std::unordered_map<std::string, GameObject*> GetBoneMapping() const { return boneMapping; }
+
+  private:
+    void SetBoneMapping();
```
method outside variables defined

SirKaza commented on SobrassadaEngine/Scene/Components/Standalone/AnimationComponent.h:
```diff
@@ -0,0 +1,52 @@
+
+    void Save(rapidjson::Value& targetState, rapidjson::Document::AllocatorType& allocator) const override;
+    void AddAnimation(UID resource);
+    UID GetAnimationResource() const { GLOG("Resource AnimUID  is: %d",resource) return resource; }
```
Not sure if this should be inline with the GLOG cause its not one line

SirKaza commented on SobrassadaEngine/Scene/Components/Standalone/AnimationComponent.h:
```diff
@@ -0,0 +1,52 @@
+#pragma once
+
+#include "Globals.h"
+#include "Scene/Components/Component.h"
```
#include "Component"

SirKaza commented on SobrassadaEngine/Scene/Components/Standalone/AnimationComponent.h:
```diff
@@ -0,0 +1,52 @@
+
+#include "Globals.h"
+#include "Scene/Components/Component.h"
+#include "AnimController.h"
```
Forward declaration of GameObject, AnimController, AnimationComponent and ResourceAnimation. All variables that are pointers or reference can be forward declared if not methods of these classes used in header

SirKaza commented on SobrassadaEngine/Scene/Components/Standalone/AnimationComponent.h:
```diff
@@ -0,0 +1,52 @@
+#include "Globals.h"
+#include "Scene/Components/Component.h"
+#include "AnimController.h"
+#include <Libs/rapidjson/document.h>
```
#include "rapidjson/document.h"

SirKaza commented on SobrassadaEngine/Scene/Components/Standalone/AnimationComponent.cpp:
```diff
@@ -0,0 +1,389 @@
+#include "AnimationComponent.h"
+#include "ResourceManagement/Resources/ResourceAnimation.h"
```
No need of ResourceManagement/Resources in all the resources includes, there is 2 more below

SirKaza commented on SobrassadaEngine/Scene/Components/Standalone/AnimationComponent.cpp:
```diff
@@ -0,0 +1,389 @@
+#include "SceneModule.h"
+
+#include "imgui.h"
+#include <Math/Quat.h>
```
""

SirKaza commented on SobrassadaEngine/Scene/Components/Standalone/AnimationComponent.cpp:
```diff
@@ -0,0 +1,389 @@
+
+AnimationComponent::~AnimationComponent()
+{
+    if (animController != nullptr)
```
no need of if and = nullptr

SirKaza commented on SobrassadaEngine/Scene/Components/Standalone/AnimationComponent.cpp:
```diff
@@ -0,0 +1,389 @@
+{
+    if (animController != nullptr && resource != 0)
+    {
+        animController->Play(resource, true);
```
All these if until OnResume can be one line IF

SirKaza commented on SobrassadaEngine/Scene/Components/Standalone/AnimController.h:
```diff
@@ -0,0 +1,41 @@
+    update_status Update(float deltaTime);  
+    void Play(UID resource, bool loop);
+    void Stop();
+    void Pause();
```
Pause and resume do it here instead of cpp, they are one line

SirKaza commented on SobrassadaEngine/Scene/Scene.cpp:
```diff
@@ -23,6 +23,9 @@
 #include "ResourceModel.h"
 #include "ResourcePrefab.h"
 #include "ResourcesModule.h"
+#include "Scene/Components/ComponentUtils.h"
```
no need of Scene/Components here and 2 below

SirKaza commented on SobrassadaEngine/UI/EngineEditors/Editor/AnimatorEditor.h:
```diff
@@ -0,0 +1,23 @@
+class GameObject;
+class ResourceAnimation;
+class AnimationComponent;
+class AnimatorEditor : public EngineEditorBase
```
If this is deleted, delete the files

DHazardFenrir commented:
I addressed all the changes requested. It should work now, even if the model doesn't have animations (there is an if that if the model doesnt have it doesn't apply the resourceModel->SetAllAnimations and SetAnimation). I also moved all asset files to the engineTests repo. 

Now for testing is moving the zombunny or player to assets, import it, load it, select the root gameObject and then press play and stop. 

The thing is, github told me I still had conflicts with vs2slice branch and that I needed to do some merge. I did it, but now my project isn't compiling because I have this linking error 'cannot open input file 'SobrassadaEngine.lib'

DHazardFenrir commented:
I ran it in release, but I had to rebuild only the sobrasadaScripts project, since if I tried to do a full rebuild the error I mentioned popped up

SirKaza reviewed (approved):
Tested in debug and release with zoombunny, works as expected.
Good job with the implementation and good use of const.

DHazardFenrir commented:
I checked and it says the branch is now up to date with vs2slice and main.

srLlobet commented:
Are the instructions old? I tried with the player, zombunny and a random animated wolf i found online, but after making a new scene, importing the gltfs, I add the animation component and there are no avaliable animations and nothing to do?

DHazardFenrir commented:
> Are the instructions old? I tried with the player, zombunny and a random animated wolf i found online, but after making a new scene, importing the gltfs, I add the animation component and there are no avaliable animations and nothing to do?

I said that if you add the component manually its it not going to work. The animation component should appear on the root when you load the models. I tested it on release and imported and loaded the zombunny again and it shows the animation component on the gameobject root.

IDarkgenesis commented on SobrassadaEngine/Scene/Components/Standalone/AnimController.cpp:
```diff
@@ -0,0 +1,190 @@
+    }
+}
+
+Quat AnimController::Interpolate(Quat& first, Quat& second, float lambda)
```
Both Quat can be const and the function aswell

IDarkgenesis reviewed (changes requested):
Works as expected, very good job, please try merge with last changes from Marc that changes how models are loaded and now they load correctly, notify on Discord as soon as the change is done and I will change the status ASAP.

👍
